DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,135 @@
+
+1. Install [Ruby](https://www.ruby-lang.org/en/documentation/installation/) 3.1.0 or later for your operating system.
+2. Install the required gems by running `bundle install`.
+3. Run `bundle exec jekyll serve` to start the Jekyll server locally. (This can take several minutes to be ready.)
+4. Open your browser and navigate to `http://localhost:4000` to view the rendered HTML pages.
+
+## Using the `spec-insert` Jekyll plugin
+The `spec-insert` Jekyll plugin is used to insert API components into the markdown files. The plugin downloads the [latest OpenSearch Specification](https://github.com/opensearch-project/opensearch-api-specification) and renders the API components from the spec. This aims to reduce the manual effort required to keep the documentation up-to-date.
+
+To use this plugin, make sure that you have installed Ruby 3.1.0 or later and the required gems by running `bundle install`.
+
+Edit your markdown file and insert the following snippet where you want an API component to be rendered:
+
+```markdown
+<!-- spec_insert_start 
+api: <API_NAME>
+component: <COMPONENT_NAME>
+other_param: <OTHER_PARAM>
+-->
+
+This is where the API component will be inserted.
+Everything between the `spec_insert_start` and `spec_insert_end` tags will be overwritten.
+
+<!-- spec_insert_end -->
+```
+
+Then run the follow Jekyll command to render the API components:
+```shell
+bundle exec jekyll spec-insert
+```
+
+If you are working on multiple markdown files and do not want to keep running the `jekyll spec-insert` command, you can add the `--watch` (or `-W`) flag to the command to watch for changes in the markdown files and automatically render the API components:
+
+```shell
+bundle exec jekyll spec-insert --watch
+```
+
+Depending on the text editor you are using to edit the markdown, you may need to manually reload the file from disk to see the changes applied by the plugin if the editor does not automatically reload the file periodically.
+
+The plugin will pull the newest OpenSearch API spec from its [repository](https://github.com/opensearch-project/opensearch-api-specification) if the spec file does not exist locally or if it is older than 24 hours. To tell the plugin to always pull the newest spec, you can add the `--refresh-spec` (or `-R`) flag to the command:
+
+```shell
+bundle exec jekyll spec-insert --refresh-spec
+```
+
+### Inserting query parameters
+
+To insert query parameters table of the, use the following snippet:
+
+```markdown
+<!-- spec_insert_start
+api: cat.indices
+component: query_parameters
+-->
+<!-- spec_insert_end -->
+```
+
+This will insert the query parameters of the `cat.indices` API into the `.md` file with three default columns: `Parameter`, `Type`, and `Description`. There are 5 columns that can be inserted: `Parameter`, `Type`, `Description`, `Required`, and `Default`. When `Required`/`Default` is not chosen, the info will be written in the `Description` column.
```

Comment:
```suggestion
This will insert the query parameters of the `cat.indices` API into the `.md` file with three default columns: `Parameter`, `Type`, and `Description`. There are five columns that can be inserted: `Parameter`, `Type`, `Description`, `Required`, and `Default`. When `Required`/`Default` is not chosen, the information will be written in the `Description` column.
```